clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2229,6 +2241,41 @@ static bool CheckResourceHandle(
 // returning an ExprError
 bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
   switch (BuiltinID) {
+  case Builtin::BI__builtin_hlsl_adduint64: {
+    if (SemaRef.checkArgCount(TheCall, 2))
+      return true;
+    if (CheckVectorElementCallArgs(&SemaRef, TheCall))
+      return true;
+    if (CheckUnsignedIntRepresentations(&SemaRef, TheCall))
+      return true;
+
+    // CheckVectorElementCallArgs(...) guarantees both args are the same type.
+    assert(TheCall->getArg(0)->getType() == TheCall->getArg(1)->getType() &&
+           "Both args must be of the same type");
+
+    // ensure both args are vectors
+    auto *VTy = TheCall->getArg(0)->getType()->getAs<VectorType>();
+    if (!VTy) {
+      SemaRef.Diag(TheCall->getBeginLoc(), diag::err_vec_builtin_non_vector)
+          << "AddUint64" << /*all*/ 1;
+      return true;
+    }
+
+    // ensure both args have 2 elements, or both args have 4 elements
+    int NumElementsArg = VTy->getNumElements();
+    if (NumElementsArg != 2 && NumElementsArg != 4) {
+      SemaRef.Diag(TheCall->getBeginLoc(),
+                   diag::err_invalid_even_odd_vector_element_count)
+          << NumElementsArg << 2 << 4 << /*even*/ 0 << /*operand*/ 1;
+      return true;
+    }
+
```

Comment:
We should also probably check that the vector element type is unsigned 32-bit integer. The requirement of size 2 or 4 is because the size must be a multiple of 64-bits. As written this looks like a `uint16_t2` would slip by without diagnostics even though there isn't enough data to perform a 64-bit add.

Reply:
Indeed, a `uint16_t2` does slip by without diagnostics.

@inbelic did suggest to check that the input integers are 32-bits, but neither of us were able to figure out how to access integer bit count from a `clang::QualType`. Perhaps you would know how it could be done?

Reply:
Nevermind. I found a way thanks to @bob80905 
```c++
// ensure arg integers are 32-bits
CharUnits CharSize = SemaRef.getASTContext().getTypeSizeInChars(VTy->getElementType());
if (CharSize.getQuantity() != 4) {
  SemaRef.Diag(...);
  return true;
}
```

Reply:
I'm a bit lost. Here in sema, you are verifying that the vec element type is 32 bits, regardless if it's 2 or 4 elements in size. However, in CGBuiltin.cpp, you are allowing for an arg to be a vector of 4 elements, and possibly constructing a 64 bit value by combining 4 16-bit parts. Which means you are throwing away the top 16 bits of each part. Is my understanding correct?
If we are confident with this assertion, then that means you're constructing 128 bit integers in CGBuiltin no?

Reply:
The 4-element vector size version of AddUint64 is the same as the 2-element vector size version of AddUint64 performed twice. 
AddUint64 returns a uint2 when its args are uint2, and AddUint64 returns a uint4 when its args are uint4.

Reply:
uint2: `AddUint64( A:{low, high}, B:{low, high} )` and the result is one pair of 32-bit integers (low, high)

uint4: `AddUint64( A:{low0, high0, low1, high1}, B:{low0, high0, low1, high1} )` is equivalent to
```
AddUint64( A:{low0, high0}, B{low0, high0} )
AddUint64( A:{low1, high1}, B{low1, high1} )
```
so the result would be two pairs of 32-bit integers, (low0, high0) and (low1, high1), stored as a uint4: (low0, high0, low1, high1)